CSE-883 config check routine

voodoods opened the pull request:
**Jira Issue**

https://8select.atlassian.net/browse/CSE-883

**Summary**

- add plugin validation install routine: prevent export cronjobs from being scheduled / executed if config is not valid
- add plugin validation frontend routine: disables plugin modules by default and shows error messages in the shop backend if something is not valid; enables the plugin module functionalities as soon as the plugin configuration is valid
- validate plugin configuration in manual exports window
- validate plugin configuration in attribute mappings window

**Info**
Options for widget position will be set by default at installation with CSE-880 and can't be set empty. Therefore a validation for these options to be empty is not necessary.


**Checklist**

- [x] PR title is prefixed with Jira Issue Id - e.g. CSE-42
- [x] Add feature / bug label
- [x] Unit tests for new / changed logic exist OR no logic changes are passing

![bildschirmfoto 2018-06-19 um 11 52 58](https://user-images.githubusercontent.com/1707307/41590471-56d0cfe4-73b7-11e8-91bd-2282e87fc1d2.png)

![bildschirmfoto 2018-06-19 um 11 52 09](https://user-images.githubusercontent.com/1707307/41590441-46295468-73b7-11e8-8ff8-27a8669b3c24.png)

akleiber commented on CseEightselectBasic/Components/GetPluginConfig.php:
```diff
@@ -0,0 +1,65 @@
+  */
+  public static function getActiveState()
+  {
+    $config = Shopware()->Container()->get('shopware.plugin.config_reader');
```
to make this more performant you should use the cached config reader like this: `$config = Shopware()->Container()->get('shopware.plugin.cached_config_reader');`

see the change here https://github.com/8select/shopware-plugin-sob/pull/18/files#diff-56b2d8a2e568308b9920cba249ed8687R116

akleiber commented on CseEightselectBasic/CseEightselectBasic.php:
```diff
@@ -416,16 +418,34 @@ private function getClasses(ModelManager $modelManager)
+     * @throws \Exception
+     */
+
+    public function cseValidatePluginConfig()
```
please move this logic inside the `GetPluginConfig` class and rename the thing to something like `ConfigValidator` to have all stuff related plugin config validation in one place

there should be a method name `isConfigValid` that just returns a boolean - this method then can be used whereever you need to check for a valid config

voodoods replied:
Good point. I think we should consider to refactor the Bootstrap file in another branch, since it is already bloated with all kinds of stuff that doesn't need to be in there (the cronjob setup, the 8select attributes definition, etc...). I think we can move a lot of it to separate classes.

Interesting read in this regard: https://developers.shopware.com/blog/2015/11/11/best-practices-of-shopware-plugin-development/

akleiber commented on CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_attribute_config/view/list/eightselect_attribute.js:
```diff
@@ -47,6 +47,123 @@ Ext.define(
       };
+    },
+
+    initComponent: function () {
```
wow this looks very complicated - I would just add a message "config invalid" with a link to the plugin configuration page

the plugin configuration page should be the central place to output help regarding the config and it's validity

those fancy growl messages are not very helpful - no customer will know what we mean with those errors without context and for sure wont remember what was wrong when he/she finally navigated to the plugin config page

voodoods commented:
@akleiber calling the plugin configuration page is currently not possible in Shopware due to a bug, see also: https://github.com/shopware/shopware/pull/1053
Unfortunately it works with every other backend application 😑

Also as far as I can see the plugin configuration files does not allow that much customization. Within the allowed items I can't find anything that would allow us to execute custom methods or display notification containers: https://github.com/shopware/shopware/blob/5.4/engine/Shopware/Components/Plugin/schema/config.xsd 

Maybe we should reconsider how to handle the error messages in the first release of this feature and do it properly as soon as its technically possible by Shopware.

voodoods commented:
@akleiber Update:

* added check routine for size definitions from attribute groups 
* made growl messages sticky so that they don't disappear by themselves
* provided link to the plugin's documentation via "Mehr Infos" button in the size definitions growl notification

Please review again 👍

voodoods commented:
@akleiber Here we go again:

- Moved plugin validation for cronjobs to `ArticleExport.php`
- Removed Growl Messages on Plugin Validation
- Show Plugin Validation Error in Modal Window upon Module Initialization

See updated screenshots in the summary 👍

rschweizer commented on CseEightselectBasic/Components/ConfigValidator.php:
```diff
@@ -0,0 +1,91 @@
+  public static function isPluginActive()
+  {
+    $config = Shopware()->Container()->get('shopware.plugin.cached_config_reader');
+    $isActive = $config->getByPluginName('CseEightselectBasic')['8s_enabled'];
```
There is a lot of code duplication here. Please make a private function to retrieve the `CseEightselectBasic` config, e.g.:
```php
private static function getCseEightselectBasicConfiguration() {
  return Shopware()->Container()->get('shopware.plugin.cached_config_reader')->getByPluginName('CseEightselectBasic')
}
```

And then simply read the individual values from the return value, e.g.:
```php
public static function isPluginActive() {
  return getCseEightselectBasicConfiguration()['8s_enabled']
}
```

rschweizer commented on CseEightselectBasic/Components/ConfigValidator.php:
```diff
@@ -0,0 +1,91 @@
+   */
+  public static function hasSizeDefinitions()
+  {
+     $sizesQuery =  'SELECT * FROM s_article_configurator_groups_attributes WHERE od_cse_eightselect_basic_is_size';
```
What is the value of `od_cse_eightselect_basic_is_size`? It might be easier to just `SELECT count(od_cse_eightselect_basic_is_size) FROM s_article_configurator_groups_attributes`, instead of querying all columns from all rows and then merely using the returned `rowCount`.

rschweizer commented on CseEightselectBasic/Components/ConfigValidator.php:
```diff
@@ -0,0 +1,91 @@
+      $apiId = $this->getApiId();
+      $feedId = $this->getFeedId();
+
+      $configIsValid = $isActive && $apiId && $feedId && strlen($apiId) == 36 && strlen($feedId) == 36;
```
Maybe we could use come sort of UUID validation here or is that not a requirement? A minimal solution would be to use some sort of RegExp, but I think it might be better to use a UUID library for proper validation. If it is necessary here, that is.

voodoods replied:
Long story short: Couldn't get the validator to work properly with the current plugin validation process. Since it is not explicitly required I created a JIRA issue for that to do it some time later: https://8select.atlassian.net/browse/CSE-1021

rschweizer commented on CseEightselectBasic/Controllers/Backend/CseEightselectBasicConfigValidation.php:
```diff
@@ -0,0 +1,62 @@
+            array_push($messages, "Plugin ist nicht aktiv");
+        }
+
+        if ( strlen($apiId) === 0 || $apiId === null || $apiId === false ) {
```
This should have another order, since `strlen` does not apply to a value of type `null` or `boolean`.
So I would first check for `null`, then for `false` and lastly for `strlen`, no?

Judging from https://github.com/8select/shopware-plugin-sob/pull/26/files#diff-dbbf697f3b4ae7662b0be95e2f8b7e4eR28, you might be able to write `!$apiId || strlen($apiId) === 0`?

voodoods replied:
Actually I can do it even shorter with just `!apiId` since PHP will validate both `null` and `0` to `false`.

rschweizer commented on CseEightselectBasic/Controllers/Backend/CseEightselectBasicConfigValidation.php:
```diff
@@ -0,0 +1,62 @@
+        $isActive = ConfigValidator::isPluginActive();
+        $apiId = ConfigValidator::getApiId();
+        $feedId = ConfigValidator::getFeedId();
+        $HtmlContainer = ConfigValidator::getHtmlContainer();
```
Starting with capital 'H' on purpose?

rschweizer commented on CseEightselectBasic/Controllers/Backend/CseEightselectBasicConfigValidation.php:
```diff
@@ -0,0 +1,62 @@
+            array_push($messages, "Die hinterlegte API ID ist ungültig");
+        }
+
+        if ( strlen($feedId) === 0 || $feedId === null || $feedId === false ) {
```
See https://github.com/8select/shopware-plugin-sob/pull/26/files#diff-dbbf697f3b4ae7662b0be95e2f8b7e4eR23

rschweizer commented on CseEightselectBasic/Controllers/Backend/CseEightselectBasicConfigValidation.php:
```diff
@@ -0,0 +1,62 @@
+            array_push($messages, "Keine API ID hinterlegt");
+        }
+
+        if ( $apiId && strlen($apiId) !== 36 ) {
```
See https://github.com/8select/shopware-plugin-sob/pull/26/files#diff-4b74a4f0ab318c76f628eae656ebb30aR87

rschweizer commented on CseEightselectBasic/Controllers/Backend/CseEightselectBasicConfigValidation.php:
```diff
@@ -0,0 +1,62 @@
+            array_push($messages, "Keine Feed ID hinterlegt");
+        }
+
+        if ( $feedId && strlen($feedId) !== 36 ) {
```
See https://github.com/8select/shopware-plugin-sob/pull/26/files#diff-4b74a4f0ab318c76f628eae656ebb30aR87

rschweizer commented on CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_attribute_config/view/list/window.js:
```diff
@@ -13,6 +14,86 @@ Ext.define(
+              }
+            ],
+            listeners: {
+              close: function() {
```
You could probably just say `close: this.close`, which would make the `me` variable unnecessary, right? If correct, applies for two other listeners as well.

voodoods replied:
I just tried it, but like ExtJS doesn't like that (neither `this.close` nor `this.close()` worked). I guess it's just the way how ExtJS handles events internally. Error that occurred:
```
Uncaught TypeError: Cannot read property 'fn' of undefined
```
Its just ExtJS things... ¯\_(ツ)_/¯

rschweizer commented on CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js:
```diff
@@ -1,99 +1,158 @@
-            statusCheck(QUICK_BTN.statusUri, QUICK_BTN.id, QUICK_BTN.textEnabled, QUICK_BTN.textDisabled, quickExportStatusCheck);
+    var FULL_BTN = {
+      id: "full-export-btn",
+      textEnabled: "Produkt Voll-Export anstoßen",
```
Vielleicht statt "anstoßen" eher "ausführen"? Klingt weniger ugs. und passt besser zu "...wird ausgeführt".

rschweizer commented on CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js:
```diff
@@ -1,99 +1,158 @@
+    };
+    var QUICK_BTN = {
+      id: "quick-export-btn",
+      textEnabled: "Produkt Schnell-Update anstoßen",
```
Vielleicht statt "anstoßen" eher "ausführen"? Klingt weniger ugs. und passt besser zu "...wird ausgeführt".

rschweizer commented on CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js:
```diff
@@ -1,99 +1,158 @@
+        success: function(response) {
+          var button = Ext.getCmp(buttonId);
+          var progress = JSON.parse(response.responseText).progress;
+          if (progress === false || progress === 100 || progress === "100") {
```
`if(progress && progress == 100)` should do the trick 😉

rschweizer commented on CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js:
```diff
@@ -1,99 +1,158 @@
-    cls: "shopware-form",
-    layout: "vbox",
+  initComponent: function() {
+    var me = this;
```
You could just use `this` within the method. Why the `me` variable?

voodoods replied:
In a nutshell this is just a best practice to prevent closure scope issues. Not always necessary, but comes in handy when dealing i.e. with store functions inside a component method. For more info see: https://stackoverflow.com/questions/13364503/javascript-me-this-why

rschweizer replied:
Yes, I know, why it is done, but I could not see the reason why it was applied here, because it was not used outside of the current function's scope. It is clearly not a problem, but I was just wondering why it was applied in this case. However, if you apply it everywhere apparently, it might make sense to stay consistent. 🙆‍♂️

rschweizer commented on CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js:
```diff
@@ -1,99 +1,158 @@
-                xtype: "button",
-                scale: "large",
-                width: "100%",
+    me.items = [
```
Put this into an `else` block on the conditional below for clarification please.

Wytrykus commented:
Btw, I like this message "Benötigen Sie Hilfe? Kontaktieren Sie uns unter ..." :-)

voodoods commented on CseEightselectBasic/Resources/views/backend/cse_eightselect_basic_manual_export/view/detail/export.js:
```diff
@@ -1,148 +1,214 @@
+        success: function(response) {
+          var button = Ext.getCmp(buttonId);
+          var progress = JSON.parse(response.responseText).progress;
+          if (progress === false || progress === 100) {
```
**Note:** This can't be evaluated as `!progress` because `progress` can also be `0` which is in JavaScript gets validated to `false` but would be a valid value here.

voodoods commented:
@Wytrykus Task 1 and 2 are done now. 

**Regarding Task 3:** Unfortunately there is no "Textblock" element or something similar that we could place inside the config.xml to view a helptext for 'required' fields. Only actual form fields are allowed, see: https://developers.shopware.com/developers-guide/plugin-configuration/ 

I guess this is because Shopware already provides a (somewhat unhelpful) solution to mark files as required, where those fields become a red underline if left empty. My best guess is that they won't allow any kind of customization inside the plugin config for the sake of security or conformity. 

We could place an actual text-input-field to display a heltpext, however I'd strongly discourage this because the user might get the impression that he has to enter some kind of text in there. We should discuss other options to provide a more user friendly solution for this in another JIRA issue.
